### PR TITLE
Validating webhook implementation

### PR DESCRIPTION
**Pull request by ichbinblau**
## Description

The PR implements a validating webhook to GMC which makes the following validations:

1. Root node existence 
2. StepName validation
3. nodeName existence
4. serviceName uniqueness

## Issues

List the issue or RFC link this PR is working on. If there is no such link, please mark it as `n/a`.

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies
N/A

## Tests
Unit test: 8096d89e6d686fb9bc880519b13fe6283ed931c5
E2E test: ed3494475a49ec8434d214a8cde13bd7ea39f64f

### CODE REVIEW

File: microservices-connector/helm/templates/deployment.yaml [hunk 1]
```diff
@@ -73,3 +73,17 @@ spec:
         {{- toYaml . | nindent 8 }}
       {{- end }}
       terminationGracePeriodSeconds: 10
+---
+apiVersion: v1
+kind: Service
```
Comment (lianhao) [inline microservices-connector/helm/templates/deployment.yaml:78]: This is not the right way to define service in helm. Please take https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/chatqna/templates/service.yaml as a reference

**Review by lianhao — CHANGES_REQUESTED** [review]

File: microservices-connector/cmd/main.go [hunk 2]
```diff
@@ -28,9 +28,15 @@ import (
 	//+kubebuilder:scaffold:imports
 )
 
+const (
+	webhookServiceName      = "gmc-validating-webhook-service"
+	webhookServiceNamespace = "system"
```
Comment (lianhao) [inline microservices-connector/cmd/main.go:33]: why we have a fixed namespace here?

File: microservices-connector/internal/controller/cert.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"time"
+)
+
+const org = "intel.com"
```
[the remark this answers is not in the document]
Comment (irisdingbj) [reply microservices-connector/internal/controller/cert.go:21]: Please double check whether we added this dependency into BOM and update go.mod go.sum for our project accordingly

File: microservices-connector/internal/controller/webhook_config.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+)
+
+func CreateOrUpdateMutatingWebhookConfiguration(caPEM *bytes.Buffer, port int32, webhookService, webhookNamespace string) error {
+	// Initializing the kube client
+	kubeconfig := os.Getenv("KUBECONFIG")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
```
Comment (irisdingbj) [inline microservices-connector/internal/controller/webhook_config.go:43]: config := ctrl.GetConfigOrDie()   please get config from gmc-controller directly

File: microservices-connector/internal/controller/webhook_config.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	resource          = "gmconnector"
+	webhookConfigName = "validating-webhook-configuration"
+)
+
+var (
+	logw           = logf.Log.WithName("WebhookConfig")
+	validatingPath = fmt.Sprintf("/validate-%s-%s-%s", strings.Replace(apiGroup, ".", "-", 2), apiVersion, resource)
+)
+
+func CreateOrUpdateMutatingWebhookConfiguration(caPEM *bytes.Buffer, port int32, webhookService, webhookNamespace string) error {
```
Comment (irisdingbj) [inline microservices-connector/internal/controller/webhook_config.go:36]: change method name

**Review by lianhao — CHANGES_REQUESTED** [review]

**Review by lianhao — COMMENTED** [review]
helm part (and only helm part) LGTM

**Comment by ichbinblau** [comment]
> helm part (and only helm part) LGTM

Thanks for the review and the help on troubleshooting!

File: microservices-connector/cmd/main.go [hunk 6]
```diff
@@ -79,8 +80,34 @@ func main() {
 		tlsOpts = append(tlsOpts, disableHTTP2)
 	}
 
+	webhookServiceName := controller.GetEnvWithDefault("SERVICE_NAME", "gmc-validating-webhook-service")
```
Comment (zhlsunshine) [inline microservices-connector/cmd/main.go:83]: Hi @ichbinblau, we may need to define a constant set for these environment variable, such as `SERVICE_NAME` and `NAMESPACE`

Comment (ichbinblau) [reply microservices-connector/cmd/main.go]: Hi @zhlsunshine . Good catch. It has been resolved. PTAL.

**Review by zhlsunshine — APPROVED** [review]
LGTM

File: microservices-connector/api/v1alpha3/validating_webhook_test.go [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := validateNames(tt.args.nodes, tt.args.fldPath); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("validateNames() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
Comment (KfreeZ) [inline microservices-connector/api/v1alpha3/validating_webhook_test.go:374]: Can we add a use case for ENDPOINT? 
for example, 
```
      - name: Embedding
        internalService:
          serviceName: embedding-svc
          config:
            endpoint: /v1/embeddings
            TEI_EMBEDDING_ENDPOINT: tei-embedding-svc
```
in this step,  TEI_EMBEDDING_ENDPOINT: tei-embedding-svc means there must be a step with service name "tei-embedding-svc" in this node.

**Comment by KfreeZ** [comment]
thanks for contributing @ichbinblau I left a comment but I think it can be added in next PR.
@irisdingbj I think unit tests are good enough for validator to cover the feature, e2e tests for validator are not necessary.

File: microservices-connector/api/v1alpha3/validating_webhook_test.go (hunk 7 above)
Comment (irisdingbj) [reply microservices-connector/api/v1alpha3/validating_webhook_test.go:374]: Agree this is a good validation case

**Comment by irisdingbj** [comment]
overall LGTM,  for e2e, we need add one case, like invalid gmc config has provided and the validaitonwebhook will check and reject the invalid config

**Comment by ichbinblau** [comment]
> overall LGTM, for e2e, we need add one case, like invalid gmc config has provided and the validaitonwebhook will check and reject the invalid config

Hi， @irisdingbj  where shall I put the e2e test? In the cicd script or in the test/e2e folder based on ginkgo?

File: microservices-connector/api/v1alpha3/validating_webhook_test.go (hunk 7 above)
Comment (ichbinblau) [reply microservices-connector/api/v1alpha3/validating_webhook_test.go:374]: Sure, I try to implement it if time permits.

**Comment by irisdingbj** [comment]
Take a look at https://github.com/opea-project/GenAIInfra/tree/main/.github/workflows/scripts/e2e

**Review by KfreeZ — APPROVED** [review]
thanks for contributing this !!!

**Pull request merged**